components/brave_wallet/browser/cardano/cardano_api_impl.cc

```diff
@@ -0,0 +1,351 @@
+                              SignDataCallback callback) {
+  auto error = CheckSelectedAccountValid();
+  if (error) {
+    std::move(callback).Run(std::nullopt, std::move(error));
+    return;
+  }
+
+  delegate_->WalletInteractionDetected();
+
+  // We now support only one address per cardano account.
+  auto supported_signing_address =
+      brave_wallet_service_->keyring_service()->GetCardanoAddress(
+          selected_account_,
+          mojom::CardanoKeyId::New(mojom::CardanoKeyRole::kExternal, 0));
+  if (!supported_signing_address ||
+      supported_signing_address->address_string != address) {
+    std::move(callback).Run(
+        std::nullopt,
+        mojom::CardanoProviderErrorBundle::New(kAPIErrorInvalidRequest,
+                                               kUnknownAddressError, nullptr));
+    return;
+  }
+
+  std::vector<uint8_t> message;
+  if (!base::HexStringToBytes(payload_hex, &message)) {
+    std::move(callback).Run(
+        std::nullopt,
+        mojom::CardanoProviderErrorBundle::New(
+            kAPIErrorInvalidRequest, WalletInternalErrorMessage(), nullptr));
+    return;
+  }
+
+  auto request = mojom::SignMessageRequest::New(
+      MakeOriginInfo(delegate_->GetOrigin()), 0, selected_account_.Clone(),
+      mojom::SignDataUnion::NewCardanoSignData(mojom::CardanoSignData::New(
+          std::string(base::as_string_view(message)))),
+      mojom::CoinType::ADA,
+      brave_wallet_service_->network_manager()->GetCurrentChainId(
+          mojom::CoinType::ADA, delegate_->GetOrigin()));
+
+  brave_wallet_service_->AddSignMessageRequest(
+      std::move(request),
+      base::BindOnce(&CardanoApiImpl::OnSignMessageRequestProcessed,
+                     weak_ptr_factory_.GetWeakPtr(), selected_account_.Clone(),
+                     std::move(supported_signing_address->payment_key_id),
+                     std::move(message), std::move(callback)));
+  delegate_->ShowPanel();
+}
+
+void CardanoApiImpl::OnSignMessageRequestProcessed(
+    const mojom::AccountIdPtr& account_id,
+    const mojom::CardanoKeyIdPtr& key_id,
+    const std::vector<uint8_t>& message,
+    SignDataCallback callback,
+    bool approved,
+    mojom::EthereumSignatureBytesPtr signature,
+    const std::optional<std::string>& error) {
+  if (error) {
+    std::move(callback).Run(std::nullopt,
+                            mojom::CardanoProviderErrorBundle::New(
+                                kAPIErrorInternalError, *error, nullptr));
+    return;
+  }
+
+  if (!approved) {
+    std::move(callback).Run(
+        std::nullopt,
+        mojom::CardanoProviderErrorBundle::New(
+            kDataSignUserDeclined,
+            l10n_util::GetStringUTF8(IDS_WALLET_USER_REJECTED_REQUEST),
+            nullptr));
+    return;
+  }
+
+  auto sig_data = brave_wallet_service_->keyring_service()
+                      ->SignCip30MessageByCardanoKeyring(
+                          selected_account_.Clone(), key_id, message);
```

Comment:
is this supposed to be `account_id.clone()` @cypt4 ?

It seems like this could be susceptible to the user changing the address while the transaction is awaiting confirmation and then getting signed by the incorrect address.

It seems like this introduces a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue